README.md
# Wikidatum

This gem supports making requests to the [new Wikidata/Wikibase REST API](https://doc.wikimedia.org/Wikibase/master/js/rest-api/).

The [Wikimedia Docs on the Wikibase REST API data format differences](https://doc.wikimedia.org/Wikibase/master/php/rest_data_format_differences.html) are also very useful for interacting with/contributing to this gem.

**The gem is currently in development. It's ready to be used, but you should be careful when making edits with it to ensure it's working correctly.** It's missing some key features, namely authentication support, but the core of the library works.

I reserve the right to make breaking changes while the library is still in the 0.x release series, although I'll avoid them unless I believe them to be significantly better for the library's usability/maintainability (or are necessary due to breaking changes in the REST API itself). If they do happen, I'll make them clear in the Changelog.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'wikidatum'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install wikidatum

## Usage

You can view the YARD docs on GitHub Pages [here](https://connorshea.github.io/wikidatum/index.html). Generally, you'll want to look at the docs for {Wikidatum::Client} to get started.

Currently, the gem is able to hit a few of the basic endpoints, and currently has no way to provide authentication. The additional features will be added later.

```ruby
require 'wikidatum'

wikidatum_client = Wikidatum::Client.new(
  user_agent: 'REPLACE ME WITH THE NAME OF YOUR BOT!',
  wikibase_url: 'https://www.wikidata.org',
  # NOTE: To edit as a bot, you need to authenticate as a user with the Bot
  # flag. If you don't have that flag on your Wikibase User, you'll get a
  # 403 error.
  bot: true
)

# Get an item from the Wikibase instance.
item = wikidatum_client.item(id: 'Q2') #=> Wikidatum::Item

# Get the statements from the item.
item.statements #=> Array<Wikidatum::Statement>

# Get the statments for property P123 on the item.
item.statements(properties: ['P123']) #=> Array<Wikidatum::Statement>

# Get all the labels for the item.
item.labels #=> Array<Wikidatum::Term>

# Get the English label for the item.
item.label(lang: :en) #=> Wikidatum::Term

# Get the actual value for the label.
item.label(lang: :en).value #=> "Earth"

# Get the values for all English aliases on this item.
item.aliases(langs: [:en]).map(&:value) #=> ["Planet Earth", "Pale Blue Dot"]

# Get all labels for a given item.
wikidatum_client.labels(id: 'Q2') #=> Array<Wikidatum::Term>

# Get a specific statement from its ID.
statement_id = 'Q123$4543523c-1d1d-1111-1e1e-11b11111b1f1'
statement = wikidatum_client.statement(id: statement_id) #=> Wikidatum::Statement

# Add a statement to Q193581 for P577 (publication date) that has a time value of November 16, 2004.
wikidatum_client.add_statement(
  id: 'Q193581',
  property: 'P577',
  value: Wikidatum::DataType::Time.new(
    time: '+2004-11-16T00:00:00Z',
    precision: 11,
    calendar_model: 'https://www.wikidata.org/entity/Q12138'
  )
)

# Delete a statement and include an edit summary.
wikidatum_client.delete_statement(
  id: 'Q123$4543523c-1d1d-1111-1e1e-11b11111b1f1',
  comment: 'Deleting this statement because it is inaccurate.'
)
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/connorshea/wikidatum.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## Unreleased
### Change

- Add redirect handling on the `item` method, will error unless `follow_redirects` is passed as true.

## 0.3.3 - 2023-01-24
### Added

- Add a `labels` method on Wikidatum::Client for getting item label data from the new `items/{id}/labels` endpoint.

### Changed

- Move error classes into their own `Wikidatum::Errors` namespace.

## 0.3.2 - 2022-12-29
### Added

- Add support for the `commonsMedia`, `url`, and `external-id` data types.
- Add `humanized_content` method to `Wikidatum::DataType::Base` class, for easier access to a Struct, string, or nil which represents the contents of the given data type.

## 0.3.1 - 2022-12-28
### Added

- Add an error message when attempting to make a bot edit without authentication, as this will always trigger a 403 error from the REST API.

### Fixed

- Fix the creation of WikibaseItem statements, the format used in the previous release was incorrect.

## 0.3.0 - 2022-12-27
### Added

- Add `allow_ip_edits` argument on `Wikidatum::Client.new`. This protects users from making IP address-exposing edits if they haven't explicitly opted-in to doing that. The argument defaults to false.
- Add code to raise errors when various types of invalid input are passed to `Wikidatum::Client#add_statement`.
- Start testing the gem on Ruby 3.2.

### Fixed

- Update the serializers and `add_statement` method to account for various changes that have been made upstream to the Wikibase REST API.

## 0.2.1 - 2022-08-13
### Fixed

- Fix a mistake that broke loading the gem.

## 0.2.0 - 2022-08-13
### Added

- Add ability to serialize most of an item response from the Beta REST API into a usable `Wikidatum::Item` instance.
  - This enables basic functionality like `client.item(id: 'Q123')` and `client.item(id: 'Q123').statements(properties: ['P123'])`.
- Add `Client#add_statement` method for creating new statements on an item.
- Add `Client#delete_statement` method for deleting statements from an item.
- Add support for reading and writing all statement types: `novalue`, `somevalue`, `string`, `time`, `quantity`, `globecoordinate`, `monolingualtext`, and `wikibase-entityid`.

### Internal

- Add YARD docs and auto-deploy them with GitHub Pages.
- Add comprehensive unit tests for the gem.

## 0.1.0 - 2022-06-20
### Added

- Initial release, nothing is really usable yet.


lib/wikidatum/client.rb
# frozen_string_literal: true

require 'faraday'
require 'faraday/net_http'

module Wikidatum
  class Client
    ITEM_REGEX = /^Q?\d+$/.freeze
    PROPERTY_REGEX = /^P?\d+$/.freeze
    STATEMENT_REGEX = /^Q?\d+\$[\w-]+$/.freeze
    VALID_RANKS = ['preferred', 'normal', 'deprecated'].freeze
    VALID_DATA_TYPES = [
      'Wikidatum::DataType::CommonsMedia',
      'Wikidatum::DataType::ExternalId',
      'Wikidatum::DataType::GlobeCoordinate',
      'Wikidatum::DataType::MonolingualText',
      'Wikidatum::DataType::NoValue',
      'Wikidatum::DataType::Quantity',
      'Wikidatum::DataType::SomeValue',
      'Wikidatum::DataType::Time',
      'Wikidatum::DataType::WikibaseItem',
      'Wikidatum::DataType::WikibaseString',
      'Wikidatum::DataType::WikibaseUrl'
    ].freeze

    # @!private
    CONTENT_DATA_TYPES = [
      'Wikidatum::DataType::CommonsMedia',
      'Wikidatum::DataType::ExternalId',
      'Wikidatum::DataType::GlobeCoordinate',
      'Wikidatum::DataType::MonolingualText',
      'Wikidatum::DataType::Quantity',
      'Wikidatum::DataType::Time',
      'Wikidatum::DataType::WikibaseItem',
      'Wikidatum::DataType::WikibaseString',
      'Wikidatum::DataType::WikibaseUrl'
    ].freeze

    # @return [String] the root URL of the Wikibase instance we want to interact
    #   with. If not provided, will default to Wikidata.
    attr_reader :wikibase_url

    # @return [Boolean] whether this client instance should identify itself
    #   as a bot when making requests.
    attr_reader :bot

    # @return [String] the UserAgent header to send with all requests to the
    #   Wikibase API.
    attr_reader :user_agent

    # @return [Boolean] whether this client should allow non-GET requests if
    #   authentication hasn't been provided. Defaults to false.
    attr_reader :allow_ip_edits

    # Create a new Wikidatum::Client to interact with the Wikibase REST API.
    #
    # @example
    #  wikidatum_client = Wikidatum::Client.new(
    #    user_agent: 'Bot Name',
    #    wikibase_url: 'https://www.wikidata.org',
    #    bot: true
    #  )
    #
    # @param user_agent [String] The UserAgent header to send with all requests
    #   to the Wikibase API. This will be prepended with the string "Wikidatum
    #   Ruby gem vX.X.X:".
    # @param wikibase_url [String] The root URL of the Wikibase instance we want
    #   to interact with. If not provided, will default to
    #   `https://www.wikidata.org`. Do not include a `/` at the end of the URL.
    # @param bot [Boolean] Whether requests sent by this client instance should
    #   be registered as bot requests. Defaults to `true`. If the user is not
    #   authenticated, or if the user is not marked with the Bot flag in the
    #   Wikibase instance, you'll be unable to make edits with this set to
    #   `true`.
    # @param allow_ip_edits [Boolean] whether this client should allow non-GET
    #   requests if authentication hasn't been provided. Defaults to false. If
    #   this is set to true, the IP address of the device from which the
    #   request was sent will be credited for the edit. Make sure not to allow
    #   these edits if you don't want your IP address (and in many cases, a
    #   very close approximation of your physical location) exposed publicly.
    # @return [Wikidatum::Client]
    def initialize(user_agent:, wikibase_url: 'https://www.wikidata.org', bot: true, allow_ip_edits: false)
      raise ArgumentError, "Wikibase URL must not end with a `/`, got #{wikibase_url.inspect}." if wikibase_url.end_with?('/')

      @user_agent = "Wikidatum Ruby gem v#{Wikidatum::VERSION}: #{user_agent}"
      @wikibase_url = wikibase_url
      @bot = bot
      @allow_ip_edits = allow_ip_edits

      Faraday.default_adapter = :net_http
    end

    # Get an item from the Wikibase API based on its QID.
    #
    # @example
    #   wikidatum_client.item(id: 'Q123')
    #   wikidatum_client.item(id: 123)
    #   wikidatum_client.item(id: '123')
    #
    # @param id [String, Integer] Either a string or integer representation of
    #   the item's QID, e.g. `"Q123"`, `"123"`, or `123`.
    # @param follow_redirects [Boolean] Whether to follow a redirect if it
    #   occurs. Will only follow the first redirect no matter what.
    # @return [Wikidatum::Item]
    def item(id:, follow_redirects: false)
      raise ArgumentError, "#{id.inspect} is an invalid Wikibase QID. Must be an integer, a string representation of an integer, or in the format 'Q123'." unless id.is_a?(Integer) || id.match?(ITEM_REGEX)

      id = coerce_item_id(id)

      response = get_request("/entities/items/#{id}", follow_redirects: follow_redirects)

      puts JSON.pretty_generate(response) if ENV['DEBUG']

      Wikidatum::Item.marshal_load(response)
    end

    # Get labels for an item from the Wikibase API based on the item's QID.
    #
    # @example
    #   wikidatum_client.labels(id: 'Q123') #=> [<Wikidatum::Term lang="en" value="Foo">, <Wikidatum::Term lang="es" value="Bar">]
    #   wikidatum_client.labels(id: 123)
    #   wikidatum_client.labels(id: '123')
    #
    # @param id [String, Integer] Either a string or integer representation of
    #   the relevant item's QID, e.g. `"Q123"`, `"123"`, or `123`.
    # @return [Array<Wikidatum::Term>] This can, theoretically, be empty if the item has no labels.
    def labels(id:)
      raise ArgumentError, "#{id.inspect} is an invalid Wikibase QID. Must be an integer, a string representation of an integer, or in the format 'Q123'." unless id.is_a?(Integer) || id.match?(ITEM_REGEX)

      id = coerce_item_id(id)

      response = get_request("/entities/items/#{id}/labels")

      puts JSON.pretty_generate(response) if ENV['DEBUG']

      response.to_a.map { |lang, val| Wikidatum::Term.new(lang: lang, value: val) }
    end

    # Get a statement from the Wikibase API based on its ID.
    #
    # @example
    #   wikidatum_client.statement(id: 'Q123$f004ec2b-4857-3b69-b370-e8124f5bd3ac')
    #
    # @param id [String] A string representation of the statement's ID.
    # @return [Wikidatum::Statement]
    def statement(id:)
      raise ArgumentError, "#{id.inspect} is an invalid Wikibase Statement ID. Must be a string in the format 'Q123$f004ec2b-4857-3b69-b370-e8124f5bd3ac'." unless id.match?(STATEMENT_REGEX)

      response = get_request("/statements/#{id}")

      puts JSON.pretty_generate(response) if ENV['DEBUG']

      Wikidatum::Statement.marshal_load(response)
    end

    # Add a statement to an item.
    #
    # NOTE: Adding references/qualifiers with `add_statement` is untested and
    # effectively unsupported for now.
    #
    # @example Add a string statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P23',
    #     value: Wikidatum::DataType::WikibaseString.new(string: 'Foo'),
    #     comment: 'Adding something or another.'
    #   )
    #
    # @example Add a 'no value' statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::NoValue.new(
    #       type: :no_value,
    #       value: nil
    #     )
    #   )
    #
    # @example Add an 'unknown value' statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::SomeValue.new(
    #       type: :some_value,
    #       value: nil
    #     )
    #   )
    #
    # @example Add a globe coordinate statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::GlobeCoordinate.new(
    #       latitude: 52.51666,
    #       longitude: 13.3833,
    #       precision: 0.01666,
    #       globe: 'https://wikidata.org/entity/Q2'
    #     )
    #   )
    #
    # @example Add a monolingual text statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::MonolingualText.new(
    #       language: 'en',
    #       text: 'Foobar'
    #     )
    #   )
    #
    # @example Add a quantity statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::Quantity.new(
    #       amount: '+12',
    #       unit: 'https://wikidata.org/entity/Q1234'
    #     )
    #   )
    #
    # @example Add a time statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::Time.new(
    #       time: '+2022-08-12T00:00:00Z',
    #       precision: 11,
    #       calendar_model: 'https://wikidata.org/entity/Q1234'
    #     )
    #   )
    #
    # @example Add a Wikibase item statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::WikibaseItem.new(
    #       id: 'Q1234'
    #     )
    #   )
    #
    # @example Add a URL statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::WikibaseUrl.new(
    #       string: 'https://example.com'
    #     )
    #   )
    #
    # @example Add an External ID statement.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::ExternalId.new(
    #       string: '123'
    #     )
    #   )
    #
    # @example Add a statement for an image, video, or audio file from Wikimedia Commons.
    #   wikidatum_client.add_statement(
    #     id: 'Q123',
    #     property: 'P124',
    #     value: Wikidatum::DataType::CommonsMedia.new(
    #       string: 'FooBar.jpg'
    #     )
    #   )
    #
    # @param id [String, Integer] the ID of the item on which the statement will be added.
    # @param property [String, Integer] property ID in the format 'P123', or an integer.
    # @param value [Wikidatum::DataType::CommonsMedia, Wikidatum::DataType::ExternalId, Wikidatum::DataType::GlobeCoordinate, Wikidatum::DataType::MonolingualText, Wikidatum::DataType::Quantity, Wikidatum::DataType::WikibaseString, Wikidatum::DataType::Time, Wikidatum::DataType::WikibaseItem, Wikidatum::DataType::WikibaseUrl, Wikidatum::DataType::NoValue, Wikidatum::DataType::SomeValue] the value of the statement being created.
    # @param qualifiers [Array<Wikidatum::Qualifier>]
    # @param references [Array<Wikidatum::Reference>]
    # @param rank [String, Symbol] Valid ranks are 'preferred', 'normal', or
    #   'deprecated'. Defaults to 'normal'. Also accepts Symbol for these ranks.
    # @param tags [Array<String>]
    # @param comment [String, nil]
    # @return [Boolean] True if the request succeeded.
    def add_statement(id:, property:, value:, qualifiers: [], references: [], rank: 'normal', tags: [], comment: nil)
      raise ArgumentError, "#{id.inspect} is an invalid Wikibase QID. Must be an integer, a string representation of an integer, or in the format 'Q123'." unless id.is_a?(Integer) || id.match?(ITEM_REGEX)
      raise ArgumentError, "#{property.inspect} is an invalid Wikibase PID. Must be an integer, a string representation of an integer, or in the format 'P123'." unless property.is_a?(Integer) || property.match?(PROPERTY_REGEX)
      raise ArgumentError, "#{rank.inspect} is an invalid rank. Must be normal, preferred, or deprecated." unless VALID_RANKS.include?(rank.to_s)
      raise ArgumentError, "Expected an instance of one of Wikidatum::DataType's subclasses for value, but got #{value.inspect}." unless VALID_DATA_TYPES.include?(value.class.to_s)

      id = coerce_item_id(id)
      property = coerce_property_id(property)

      case value.class.to_s
      when 'Wikidatum::DataType::NoValue'
        statement_hash = {
          property: {
            id: property
          },
          value: {
            type: 'novalue'
          }
        }
      when 'Wikidatum::DataType::SomeValue'
        statement_hash = {
          property: {
            id: property
          },
          value: {
            type: 'somevalue'
          }
        }
      when *CONTENT_DATA_TYPES
        statement_hash = {
          property: {
            id: property
          },
          value: {
            type: 'value',
            content: value.marshal_dump
          }
        }
      end

      body = {
        statement: statement_hash.merge(
          {
            qualifiers: qualifiers,
            references: references,
            rank: rank.to_s
          }
        )
      }

      response = post_request("/entities/items/#{id}/statements", body, tags: tags, comment: comment)

      puts JSON.pretty_generate(response) if ENV['DEBUG']

      response.success?
    end

    # Delete a statement from an item.
    #
    # @example
    #   wikidatum_client.delete_statement(
    #     id: 'Q123$4543523c-1d1d-1111-1e1e-11b11111b1f1',
    #     comment: "Deleting this statement because it's bad."
    #   )
    #
    # @param id [String] the ID of the statemnt being deleted.
    # @param tags [Array<String>]
    # @param comment [String, nil]
    # @return [Boolean] True if the request succeeded.
    def delete_statement(id:, tags: [], comment: nil)
      raise ArgumentError, "#{id.inspect} is an invalid Wikibase Statement ID. Must be a string in the format 'Q123$f004ec2b-4857-3b69-b370-e8124f5bd3ac'." unless id.match?(STATEMENT_REGEX)

      response = delete_request("/statements/#{id}", tags: tags, comment: comment)

      puts JSON.pretty_generate(response) if ENV['DEBUG']

      response.success?
    end

    # Is the current instance of Client authenticated as a Wikibase user?
    #
    # @return [Boolean]
    def authenticated?
      # TODO: Make it possible for this to be true once authentication
      #   is implemented.
      false
    end

    # Does the current instance of Client allow anonymous IP-based edits?
    #
    # @return [Boolean]
    def allow_ip_edits?
      @allow_ip_edits
    end

    # Is the current instance of Client editing as a bot?
    #
    # @return [Boolean]
    def bot?
      @bot
    end

    private

    # For now this just returns the `@wikibase_url`, but in the future the API
    # routes will presumably be nested further, so this is just future-proofing
    # to allow that to be easily changed later.
    #
    # @return [String] URL for the Wikibase API endpoint.
    def api_url
      @api_url ||= "#{@wikibase_url}/w/rest.php/wikibase/v0"
    end

    # Default headers to be sent with every request.
    #
    # @return [Hash] A hash of some headers that should be used when sending a request.
    def universal_headers
      @universal_headers ||= {
        'User-Agent' => @user_agent,
        'Content-Type' => 'application/json'
      }
    end

    # Make a GET request to a given Wikibase endpoint.
    #
    # @param path [String] The relative path for the API endpoint.
    # @param params [Hash] Query parameters to send with the request, if any.
    # @param follow_redirects [Boolean] Whether to follow a redirect if it occurs. Will only follow the first redirect no matter what.
    # @return [Hash] JSON response, parsed into a hash.
    def get_request(path, params = nil, follow_redirects: false)
      url = "#{api_url}#{path}"

      response = Faraday.get(url, params, universal_headers)

      # If we have a permanent redirect response and the method is set to follow redirects, follow redirects.
      # Error handling if it doesn't return a 200
      if response.status == 308
        return get_request(response.headers['location'].gsub(api_url, ''), params, follow_redirects: false) if follow_redirects

        redirect_url = response.headers['location']
        raise Errors::RequestRedirectedError.new(redirect_url)
      elsif !response.success?
        puts 'Something went wrong with this request!'
        puts "Status Code: #{response.status}"
        puts response.body.inspect
      end

      # Raise an error if the response was ''.
      raise Errors::InvalidResponseError if response.body.empty?

      JSON.parse(response.body)
    end

    # Make a POST request to a given Wikibase endpoint.
    #
    # @param path [String] The relative path for the API endpoint.
    # @param body [Hash] The body to post to the endpoint.
    # @param tags [Array<String>] The tags to apply to the edit being made by this request, for PUT/POST/DELETE requests.
    # @param comment [String] The edit description, for PUT/POST/DELETE requests.
    # @return [Hash] JSON response, parsed into a hash.
    def post_request(path, body = {}, tags: [], comment: nil)
      ensure_edit_permitted!

      url = "#{api_url}#{path}"

      body[:bot] = @bot
      body[:tags] = tags unless tags.empty?
      body[:comment] = comment unless comment.nil?

      response = Faraday.post(url) do |req|
        req.body = JSON.generate(body)
        req.headers = universal_headers
      end

      puts response.body.inspect if ENV['DEBUG']

      # Error handling if it doesn't return a 200
      unless response.success?
        puts 'Something went wrong with this request!'
        puts "Status Code: #{response.status}"
        puts response.body.inspect
      end

      response
    end

    # Make a DELETE request to a given Wikibase endpoint.
    #
    # @param path [String] The relative path for the API endpoint.
    # @param tags [Array<String>] The tags to apply to the edit being made by this request, for PUT/POST/DELETE requests.
    # @param comment [String] The edit description, for PUT/POST/DELETE requests.
    # @return [Hash] JSON response, parsed into a hash.
    def delete_request(path, tags: [], comment: nil)
      ensure_edit_permitted!

      url = "#{api_url}#{path}"

      body = {}
      body[:bot] = @bot
      body[:tags] = tags unless tags.empty?
      body[:comment] = comment unless comment.nil?

      response = Faraday.delete(url) do |req|
        req.body = JSON.generate(body)
        req.headers = universal_headers
      end

      puts response.body.inspect if ENV['DEBUG']

      # Error handling if it doesn't return a 200
      unless response.success?
        puts 'Something went wrong with this request!'
        puts "Status Code: #{response.status}"
        puts response.body.inspect
      end

      response
    end

    # Coerce an Item ID in the formats 'Q123', '123' or 123 into a consistent
    # 'Q123' format. We need to have the ID in the format 'Q123' for the API
    # request, which is why coercion is necessary.
    #
    # @param id [String, Integer]
    # @return [String]
    def coerce_item_id(id)
      return id if id.to_s.start_with?('Q')

      "Q#{id}"
    end

    # Coerce a Property ID in the formats 'P123', '123' or 123 into a consistent
    # 'P123' format. We need to have the ID in the format 'P123' for the API
    # request, which is why coercion is necessary.
    #
    # @param property_id [String, Integer]
    # @return [String]
    def coerce_property_id(property_id)
      return property_id if property_id.to_s.start_with?('P')

      "P#{property_id}"
    end

    # Check if authentication has been provided, and then check if IP edits
    # are allowed. If neither condition returns true, raise an error.
    #
    # Also check if the user is performing IP edits as a bot, which will
    # always return a 403 error from the REST API, and return a specific
    # error message if so.
    #
    # @return [void]
    # @raise [Errors::DisallowedIpEditError, Errors::DisallowedBotEditError]
    def ensure_edit_permitted!
      raise Errors::DisallowedIpEditError if !authenticated? && !allow_ip_edits?
      raise Errors::DisallowedBotEditError if !authenticated? && bot?
    end
  end
end
